compiler_gym/util/minimize_trajectory.py

```diff
@@ -0,0 +1,317 @@
+        logging.debug(
+            "Bisect step=%d, left=%d, right=%d, mid=%d", step, left, right, mid
+        )
+
+        actions = all_actions[mid:] if reverse else all_actions[:mid]
+        if apply_and_test(actions):
+            logging.info(
+                "🟢 Hypothesis holds at num_actions=%d, remaining bisect steps=%d",
+                mid,
+                remaining_steps,
+            )
+            yield env
+            if reverse:
+                left = mid + 1
+            else:
+                right = mid - 1
+        else:
+            logging.info(
+                "🔴 Hypothesis does not hold at num_actions=%d, remaining bisect steps=%d",
+                mid,
+                remaining_steps,
+            )
+            if reverse:
+                right = mid - 1
+            else:
+                left = mid + 1
+
+    mid = max(left, right) - 1 if reverse else min(left, right) + 1
+    if (reverse and mid < 0) or (not reverse and mid >= len(all_actions)):
+        actions = all_actions
+        logging.info("Failed to reduce trajectory length using bisection")
+    else:
+        actions = all_actions[mid:] if reverse else all_actions[:mid]
+        logging.info(
+            "Determined that action %d of %d is the first at which the hypothesis holds: %s",
+            mid,
+            len(all_actions),
+            env.action_space.flags[all_actions[mid]],
+        )
+
+    if not apply_and_test(actions):
+        raise MinimizationError("Post-bisect sanity check failed!")
+
+    yield env
+
+
+def random_minimization(
+    env: CompilerEnv,
+    hypothesis: Hypothesis = environment_validation_fails,
+    num_steps_ratio_multiplier: float = 5,
+    init_discard_ratio: float = 0.75,
+    discard_ratio_decay: float = 0.75,
+    flakiness: int = 1,
+) -> Iterable[CompilerEnv]:
+    """Run an iterative process of randomly removing actions to minimize a
+    trajectory.
+
+    For each round of minimization, a number of actions are discarded randomly
+    and the hypothesis is tested. If the hypothesis still holds with those
+    actions removed, the minimization proceeds. Else the actions are re-inserted
+    into the trajectory and a new set of actions are removed. After a failure
+
+    Performs up to O(num_steps_ratio_multiplier * log n) evaluation rounds,
+    where n is the length of the trajectory.
+
+    :param env: An environment whose action trajectory should be minimized.
+    :param hypothesis: The hypothesis that is used to determine if a trajectory
+        is valid. A callback that accepts as argument the :code:`env:`
+        instance and returns true if the hypothesis holds, else false. The
+        hypothesis must hold on the initial trajectory.
+    :param num_steps_ratio_multiplier: A multiplier for the number of rounds of
+        minimization to perform, using log(n) the length of the trajectory as
+        the factor.
+    :param init_discard_ratio: The number of actions that will be randomly
```

Comment:
Maybe describe the exponential decay thing in the intro?

Reply:
Actually, it is not easy to understand exactly what this code does from the doc. E.g. the `init_discard_ratio` is reset on success. The process that will be undertaken is not very clear, which makes it had to know what parameter values might be sensible.